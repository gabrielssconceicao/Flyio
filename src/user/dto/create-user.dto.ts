import { ApiProperty } from '@nestjs/swagger';
import {
  IsEmail,
  IsNotEmpty,
  IsOptional,
  IsString,
  MaxLength,
  MinLength,
} from 'class-validator';
<<<<<<< HEAD
=======
import {
  dtoErrorMessages,
  DtoErrorType,
} from '../../common/utils/dto-error-messages';

>>>>>>> images
export class CreateUserDto {
  @ApiProperty({
    example: 'John Doe',
    description: 'User name',
    minLength: 3,
<<<<<<< HEAD
    maxLength: 255,
  })
  @IsString()
  @IsNotEmpty()
  @MinLength(3)
  @MaxLength(255)
  name: string;

  @ApiProperty({
    example: 'jdoe@email.com',
    description: 'User email',
  })
  @IsEmail()
  email: string;

  @ApiProperty({
    example: '123456',
    description: 'User password',
    minLength: 6,
    maxLength: 255,
  })
  @IsString()
  @IsNotEmpty()
  @MinLength(6)
  @MaxLength(255)
  password: string;
=======
    maxLength: 100,
  })
  @IsNotEmpty({ message: dtoErrorMessages('Name', DtoErrorType.REQUIRED) })
  @IsString({ message: dtoErrorMessages('Name', DtoErrorType.IS_STRING) })
  @MinLength(3, {
    message: dtoErrorMessages('Name', DtoErrorType.MIN_LENGTH, 3),
  })
  @MaxLength(100, {
    message: dtoErrorMessages('Name', DtoErrorType.MAX_LENGTH, 100),
  })
  name: string;

  @ApiProperty({
    example: 'jDoe453',
    description: 'User username',
    minLength: 3,
    maxLength: 100,
  })
  @IsNotEmpty({ message: dtoErrorMessages('Username', DtoErrorType.REQUIRED) })
  @IsString({ message: dtoErrorMessages('Username', DtoErrorType.IS_STRING) })
  @MinLength(3, {
    message: dtoErrorMessages('Username', DtoErrorType.MIN_LENGTH, 3),
  })
  @MaxLength(100, {
    message: dtoErrorMessages('Username', DtoErrorType.MAX_LENGTH, 100),
  })
  username: string;

  @ApiProperty({
    example: 'jdoe@me.com',
    description: 'User email',
  })
  @IsEmail()
  @IsNotEmpty({ message: dtoErrorMessages('Email', DtoErrorType.REQUIRED) })
  email: string;

  @ApiProperty({
    example: 'exampl3P@ssw0rd',
    description: 'User password',
    minLength: 6,
    maxLength: 100,
  })
  @IsNotEmpty({ message: dtoErrorMessages('Password', DtoErrorType.REQUIRED) })
  @IsString({ message: dtoErrorMessages('Password', DtoErrorType.IS_STRING) })
  @MinLength(6, {
    message: dtoErrorMessages('Password', DtoErrorType.MIN_LENGTH, 6),
  })
  @MaxLength(100, {
    message: dtoErrorMessages('Password', DtoErrorType.MAX_LENGTH, 100),
  })
  password: string; // change prisma to passwordHash
>>>>>>> images

  @ApiProperty({
    example: 'This is my bio',
    description: 'User bio',
    maxLength: 255,
<<<<<<< HEAD
  })
  @IsOptional()
  @IsString()
  @MaxLength(255)
  bio?: string;

  @ApiProperty({
    example: 'example.png',
    description: 'User profile image',
  })
  @IsOptional()
  @IsString()
  profileImg?: string;
=======
    required: false,
  })
  @IsOptional()
  @IsString({ message: dtoErrorMessages('Bio', DtoErrorType.IS_STRING) })
  @MaxLength(255, {
    message: dtoErrorMessages('Bio', DtoErrorType.MAX_LENGTH, 255),
  })
  bio?: string;
>>>>>>> images
}
