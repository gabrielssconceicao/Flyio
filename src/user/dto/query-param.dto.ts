import { Type } from 'class-transformer';
import { IsOptional, IsString, Max, Min } from 'class-validator';

export class QueryParamDto {
  @IsOptional()
  @IsString()
  search?: string;
  @IsOptional()
  @Min(1)
  @Max(50)
  @Type(() => Number)
  limit?: number;
  @IsOptional()
  @Min(0)
  @Type(() => Number)
  offset?: number;
}
