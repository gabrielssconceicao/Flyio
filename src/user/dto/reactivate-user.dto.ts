import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty, IsString } from 'class-validator';
import {
  dtoErrorMessages,
  DtoErrorType,
} from '../../common/utils/dto-error-messages';

export class ReactivateUserDto {
  @ApiProperty({
    example: 'krcio48hgGYIFYJfuyfJYAYu',
    description: 'Activation token',
  })
  @IsNotEmpty({ message: dtoErrorMessages('Token', DtoErrorType.REQUIRED) })
  @IsString()
  token: string;
}
