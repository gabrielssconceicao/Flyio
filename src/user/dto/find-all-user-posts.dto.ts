import { ApiProperty, OmitType } from '@nestjs/swagger';
import { FindAllResponseDto } from 'src/common/dto/find-all-response.dto';
import { PostEntity } from '../../post/entities/post.entity';

export class FindAllUsersPostsResponseDto extends FindAllResponseDto<
  Omit<PostEntity, 'user'>
> {
  @ApiProperty({
    type: [OmitType(PostEntity, ['user'])],
    description: 'List of posts of a user with limited details',
  })
  items: Omit<PostEntity, 'user'>[];
}
