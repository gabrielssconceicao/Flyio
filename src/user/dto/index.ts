import { CreateUserDto } from './create-user.dto';
import { FindAllUsersResponseDto } from './find-all-users.dto';
import { QueryParamDto } from './query-param.dto';
import { ReactivateUserDto } from './reactivate-user.dto';
import { UpdateUserDto } from './update-user.dto';

export {
  CreateUserDto,
  FindAllUsersResponseDto,
  QueryParamDto,
  ReactivateUserDto,
  UpdateUserDto,
};
