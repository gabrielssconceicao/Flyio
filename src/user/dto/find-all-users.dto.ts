import { ApiProperty } from '@nestjs/swagger';
import { PickType } from '@nestjs/mapped-types';
import { User } from '../entities/user.entity';
import { FindAllResponseDto } from '../../common/dto/find-all-response.dto';

class UserSummaryDto extends PickType(User, [
  'id',
  'name',
  'username',
  'profileImg',
] as const) {
  @ApiProperty({
    example: 'd-43-5df-df4',
    description: 'Unique identifier of the user',
  })
  id: string;

  @ApiProperty({ example: 'John Doe', description: 'Name of the user' })
  name: string;

  @ApiProperty({ example: 'jdoe', description: 'Username of the user' })
  username: string;

  @ApiProperty({
    example: 'https://example.com/profile.jpg',
    description: 'Profile image URL',
    nullable: true,
  })
  profileImg: string | null;
}

export class FindAllUsersResponseDto extends FindAllResponseDto<UserSummaryDto> {
  @ApiProperty({
    type: [UserSummaryDto],
    description: 'List of users with limited details',
  })
  items: UserSummaryDto[];
}
