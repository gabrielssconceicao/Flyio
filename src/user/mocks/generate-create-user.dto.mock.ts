import { CreateUserDto } from '../dto';
import { userMock } from './user.mock';
export const generateCreateUserDtoMock = (
  bio: boolean = false,
): CreateUserDto => {
  const dto: CreateUserDto = {
    name: userMock.name,
    username: userMock.username,
    email: userMock.email,
    password: userMock.password,
  };

  if (bio) dto.bio = userMock.bio;
  return dto;
};
