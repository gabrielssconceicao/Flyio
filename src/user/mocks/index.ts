import { generateUserMock } from './user.mock';
import { generateFindAllUsersResponseDtoMock } from './generate-find-all-users-response.dto.mock';
import { generateCreateUserDtoMock } from './generate-create-user.dto.mock';

export {
  generateUserMock,
  generateFindAllUsersResponseDtoMock,
  generateCreateUserDtoMock,
};
