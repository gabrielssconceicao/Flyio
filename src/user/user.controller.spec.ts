import { Test, TestingModule } from '@nestjs/testing';
import { JwtService } from '@nestjs/jwt';
import { UserController } from './user.controller';
import { UserService } from './user.service';
import { User } from './entities/user.entity';
import jwtConfig from '../auth/config/jwt.config';
import {
  jwtServiceMock,
  jwtConfigurationMock,
  generateTokenPayloadDtoMock,
} from '../auth/mocks';
import { TokenPayloadDto } from '../auth/dto';
import { generateFileMock } from '../cloudinary/mocks';

import { CreateUserDto, FindAllUsersResponseDto } from './dto';
import {
  generateCreateUserDtoMock,
  generateFindAllUsersResponseDtoMock,
  generateUserMock,
} from './mocks';

describe('UserController', () => {
  let controller: UserController;
  let service: UserService;

  //mocks
  let createUserDto: CreateUserDto;
  let file: Express.Multer.File;
  let user: User;
  let users: FindAllUsersResponseDto;
  let username: string;
  let id: string;
  let tokenPayload: TokenPayloadDto;
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [UserController],
      providers: [
        {
          provide: UserService,
          useValue: {
            create: jest.fn(),
            findAll: jest.fn(),
            findOne: jest.fn(),
            update: jest.fn(),
            deactivate: jest.fn(),
            removeProfilePicture: jest.fn(),
          },
        },

        {
          provide: JwtService,
          useValue: jwtServiceMock,
        },
        {
          provide: jwtConfig.KEY,
          useValue: jwtConfigurationMock,
        },
      ],
    }).compile();

    controller = module.get<UserController>(UserController);
    service = module.get<UserService>(UserService);

    // mocks
    createUserDto = generateCreateUserDtoMock();
    file = generateFileMock();
    user = generateUserMock();
    users = generateFindAllUsersResponseDtoMock();

    username = user.username;
    id = user.id;
    tokenPayload = generateTokenPayloadDtoMock();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
    expect(service).toBeDefined();
  });

  describe('<Create/ >', () => {
    it('should create a user', async () => {
      jest.spyOn(service, 'create').mockResolvedValue(user);

      const result = await controller.create(createUserDto, file);
      expect(service.create).toHaveBeenCalledWith(createUserDto, file);
      expect(result).toEqual(user);
      expect(result).toMatchSnapshot();
    });
  });

  describe('<FindAll />', () => {
    it('should return an array of users if user is logged in and token is valid', async () => {
      const query = {
        search: 'jD',
        limit: 1,
        offset: 0,
      };
      jest.spyOn(service, 'findAll').mockResolvedValue(users);

      const result = await controller.findAll(query);

      expect(service.findAll).toHaveBeenCalledWith(query);
      expect(result).toEqual(users);
      expect(result).toMatchSnapshot();
    });
  });

  describe('<FindOne />', () => {
    it('should return a user successfully', async () => {
      jest.spyOn(service, 'findOne').mockResolvedValue(user);
      const result = await controller.findOne(username);
      expect(service.findOne).toHaveBeenCalledWith(username);
      expect(result).toEqual(user);
      expect(result).toMatchSnapshot();
    });
  });

  describe('<Update />', () => {
    it('should update a user successfully', async () => {
      const updateUserDto = {
        name: 'John Doe',
        bio: 'Update bio',
      };

      const updatedUser = { ...user, ...updateUserDto };
      jest.spyOn(service, 'update').mockResolvedValue(updatedUser);
      const result = await controller.update(
        id,
        updateUserDto,
        file,
        tokenPayload,
      );

      expect(service.update).toHaveBeenCalledWith(
        id,
        updateUserDto,
        file,
        tokenPayload,
      );
      expect(result).toEqual(updatedUser);
      expect(result).toMatchSnapshot();
    });
  });

  describe('<Deactivate />', () => {
    it('should delete a user successfully', async () => {
      const message = 'User desactivated successfully';
      jest.spyOn(service, 'deactivate').mockResolvedValue({ message });
      const result = await controller.deactivate(id, tokenPayload);
      expect(service.deactivate).toHaveBeenCalledWith(id, tokenPayload);
      expect(result).toEqual({ message });
      expect(result).toMatchSnapshot();
    });
  });

  describe('<RemoveProfilePicture />', () => {
    it('should remove profile picture successfully', async () => {
      user = { ...user, profileImg: null };
      jest.spyOn(service, 'removeProfilePicture').mockResolvedValue(user);
      const result = await controller.removeProfileImg(username, tokenPayload);
      expect(service.removeProfilePicture).toHaveBeenCalledWith(
        username,
        tokenPayload,
      );
      expect(result).toEqual(user);
      expect(result).toMatchSnapshot();
    });
  });
});
