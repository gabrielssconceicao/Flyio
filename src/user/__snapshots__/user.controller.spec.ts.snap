// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController <Create/ > should create a user 1`] = `
{
  "active": true,
  "bio": "This is my bio",
  "email": "jdoe@me.com",
  "id": "42-d-f-df4",
  "name": "John Doe",
  "profileImg": "https://example.com/profile.jpg",
  "username": "jDoe453",
}
`;

exports[`UserController <Create/ > should throw ConflictException when email or username is already in use 1`] = `[ConflictException: This email or username is already associated with an existing account]`;

exports[`UserController <Delete /> should delete a user successfully 1`] = `
{
  "message": "User deleted successfully",
}
`;

exports[`UserController <Delete /> should throw an NotFoundException 1`] = `[NotFoundException: User not found]`;

exports[`UserController <FindAll /> should return an array of users 1`] = `
{
  "count": 1,
  "users": [
    {
      "id": "42-d-f-df4",
      "name": "John Doe",
      "profileImg": "https://example.com/profile.jpg",
      "username": "jDoe453",
    },
  ],
}
`;

exports[`UserController <FindOne /> should return a user successfully 1`] = `
{
  "active": true,
  "bio": "This is my bio",
  "email": "jdoe@me.com",
  "id": "42-d-f-df4",
  "name": "John Doe",
  "profileImg": "https://example.com/profile.jpg",
  "username": "jDoe453",
}
`;

exports[`UserController <FindOne /> should throw an NotFoundException 1`] = `[NotFoundException: User not found]`;

exports[`UserController <Update /> should throw an ConflictException 1`] = `[ConflictException: Email is already associated with an existing account]`;

exports[`UserController <Update /> should throw an NotFoundException 1`] = `[NotFoundException: User not found]`;

exports[`UserController <Update /> should update a user successfully 1`] = `
{
  "active": true,
  "bio": "Update bio",
  "email": "jdoe@me.com",
  "id": "42-d-f-df4",
  "name": "John Doe",
  "profileImg": "https://example.com/profile.jpg",
  "username": "jDoe453",
}
`;
